src/sundials_matrix_wrapper.cpp

```diff
@@ -453,47 +431,120 @@ void SUNMatrixWrapper::sparse_multiply(SUNMatrixWrapper *C,
         }
         for (p = Bp[j]; p < Bp[j+1]; p++)
         {
-            nnz = scatter(Bi[p], Bx[p], w.data(), x.data(), j+1, C, nnz);
+            nnz = scatter(Bi[p], Bx[p], w.data(), gsl::make_span(x), j+1, C,
+                          nnz);
             assert(nnz - Cp[j] <= m);
         }
         for (p = Cp[j]; p < nnz; p++)
             Cx[p] = x[Ci[p]]; // copy data to C
     }
     Cp[n] = nnz;
+    assert(nnz <= C->capacity());
     /*
      * do not reallocate here since we rather keep a matrix that is a bit
      * bigger than repeatedly resizing this matrix.
      */
 }
 
+void SUNMatrixWrapper::sparse_add(SUNMatrixWrapper *A, realtype alpha,
+                                  SUNMatrixWrapper *B, realtype beta) {
+    // matrix_ == nullptr is allowed on the first call
+    if (!A->matrix_ || !B->matrix_)
```

Comment:
Must check for nullptr before dereferencing (A and B, not only their matrix_ member). Pass A,B as reference?